feat: Add callback for configuring Vary header

gl-jkeys opened the pull request:
Fixes #317 

This PR adds a callback (`shouldSetVaryHeader`) to the options which is used to determine when and how to configure the Vary header.

To preserve backward compatibility with all existing `cors` usages, this callback returns `true` by default.

If this PR is accepted, I will add documentation to the README on how to use this PR's functionality.

gl-jkeys committed:
feat: Add callback for configuring Vary header

UlisesGascon commented:
I enabled the CI to run in this PR :+1:

jonchurch commented on test/test.js:
```diff
@@ -150,6 +150,198 @@ var util = require('util')
+              return true
+            }
+
+            if (req.originalUrl.startsWith('/images') && header === 'Origin') {
```
Node `0.10` failures are due to lack of `.startsWith` this language feature came in with ES6 in Node 6

joeyguerra replied:
`if (req.originalUrl.indexOf('/images') === 0 && header === 'Origin') {`

indexOf should work.

gl-jkeys replied:
Made this change!

gl-jkeys committed:
PR feedback: use indexOf, per @joeyguerra

IamLizu commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "cors",
   "description": "Node.js CORS middleware",
-  "version": "2.8.5",
+  "version": "2.9.0",
```
I think this version change is supposed to decided during release, not now. Please revert this change.

gl-jkeys replied:
done, thanks!

IamLizu commented:
Hey @gl-jkeys 👋, first of all, thank you for being an awesome contributor. 

> If this PR is accepted, I will add documentation to the README on how to use this PR's functionality.

It makes more sense if that README change is also included in this PR. Could you please take care of that?

(Note to self: I was unable to run the tests due to my own PC issue, I will check the tests again sometime later soon).

gl-jkeys committed:
revert version change

gl-jkeys committed:
Make shouldSetVaryHeader do nothing for OPTIONS requests; update tests to match; update README

gl-jkeys committed:
simplify tests to match simplified implementation

gl-jkeys commented:
Hey @IamLizu, I made the changes you requested; I also modified the PR to unconditionally set the Vary: Origin header for OPTIONS requests. (It seems unlikely, in hindsight since this PR was opened, that OPTIONS requests would be cached without respect to Origin.) I also simplified the tests to match the updated implementation.

gl-jkeys committed:
shouldSetVaryHeader -> shouldSetVaryOriginHeader

IamLizu commented:
Hey @gl-jkeys sorry I missed this one.

I will run the tests in my pc and get back to you.

IamLizu reviewed (approved):
LGTM :+1: 

I am in favor of having the option to not set _Vary_ header where it could help avoiding unncessary cache fragmentation.
